pkg/ingester/ingester.go

```diff
@@ -3406,3 +3471,103 @@ type utilizationBasedLimiter interface {
 
 	LimitingReason() string
 }
+
+var ownedSeriesRingOp = ring.WriteNoExtend // Distributor uses WriteNoExtend.
+
+// This is Starting function for ownedSeries service. Service is only started after all TSDBs are opened.
+// Pushes are not allowed yet when this function runs.
+func (i *Ingester) ownedSeriesStarting(ctx context.Context) error {
+	err := ring.WaitInstanceState(ctx, i.ingestersRing, i.lifecycler.ID, ring.ACTIVE)
+	if err != nil {
+		return err
+	}
+
+	// Compute ownedSeries before we start (and allow pushes).
+	rs, err := i.ingestersRing.GetAllHealthy(ownedSeriesRingOp)
+	if err != nil {
+		return fmt.Errorf("owned series: can't cache replica set: %w", err)
+	}
+
+	i.ownedSeriesLastReplicaSet = rs
+	// Technically ring changed, but all TSDBs at this point also have "new user" trigger set anyway.
+	i.ownedSeriesUpdate(ctx, true)
+	return nil
+}
+
+// This function runs periodically. It checks if ring has changed, and updates number of owned series for any
+// user that requires it (due to ring change, compaction, shard size change, ...).
+func (i *Ingester) ownedSeriesIter(ctx context.Context) error {
+	rs, err := i.ingestersRing.GetAllHealthy(ownedSeriesRingOp)
+	if err != nil {
+		level.Error(i.logger).Log("msg", "owned series: can't check ring for updates", "err", err)
+		return nil // If we returned error, OwnedSeries service would stop.
+	}
+
+	ringChanged := ring.HasReplicationSetChanged(i.ownedSeriesLastReplicaSet, rs)
+	i.ownedSeriesLastReplicaSet = rs
+
+	start := time.Now()
+	i.ownedSeriesUpdate(ctx, ringChanged)
+	level.Info(i.logger).Log("msg", "owned series: updated owned series for all users", "duration", time.Since(start), "ringChanged", ringChanged)
```

Comment:
[nit] This log will be quite frequent. What if we return from `ownedSeriesUpdate()` the number of tenants for which owned series has been really updated, and log this line only when `> 0` (adding a field to specify the number of updated tenants)?

Reply:
Done.